### PR TITLE
Add search command

### LINKED ISSUE #7: As a user I should be able to search for books
so that I find new books to read

### PR DESCRIPTION
**Pull request by takuyakanbr**
* Add search command
* Modify UI, Model, and Logic to support displaying search results
* Add unit, integration, and system tests for search command
* Add tests for clear, delete, select commands to verify behavior when displaying search results
* Add search command to user guide

Fixes #7

### CODE REVIEW

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 1]
```diff
@@ -44,6 +57,23 @@ public void execute_validIndexUnfilteredList_success() throws Exception {
...
+    public void execute_invalidActiveListType_throwsCommandException() throws Exception {
+        model.setActiveListType(ActiveListType.SEARCH_RESULTS);
+        DeleteCommand deleteCommand = prepareCommand(Index.fromZeroBased(0));
+
+        assertCommandFailure(deleteCommand, model, DeleteCommand.MESSAGE_WRONG_ACTIVE_LIST);
+    }
+
+    @Test
+    public void execute_invalidBook_throwsAssertionError() throws Exception {
+        DeleteCommand deleteCommand = prepareCommand(Index.fromZeroBased(0));
```
Comment (qiu-siqi) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java]: INDEX_FIRST_BOOK

File: src/test/java/seedu/address/logic/commands/SearchCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+        Model expectedModel = new ModelManager(new BookShelf(model.getBookShelf()), new UserPrefs());
+
+        assertCommandSuccess(command, model, SearchCommand.MESSAGE_SEARCHING, expectedModel);
+    }
+
+    @Test
+    public void execute_noFieldSpecifiedNoSearchTerm_success() {
+        SearchDescriptor searchDescriptor = new SearchDescriptorBuilder().build();
+        SearchCommand command = prepareCommand(searchDescriptor);
```
Comment (qiu-siqi) [inline src/test/java/seedu/address/logic/commands/SearchCommandTest.java]: Expect NullPointerException?

File: src/test/java/seedu/address/logic/commands/SelectCommandTest.java [hunk 3]
```diff
@@ -49,21 +54,43 @@ public void execute_validIndexUnfilteredList_success() {
...
+        model = new ModelManager();
+        model.setActiveListType(ActiveListType.SEARCH_RESULTS);
+        BookShelf bookShelf = new BookShelf();
+        bookShelf.addBook(TypicalBooks.ARTEMIS);
+        model.updateSearchResults(bookShelf);
+        assertExecutionFailure(INDEX_SECOND_BOOK, Messages.MESSAGE_INVALID_BOOK_DISPLAYED_INDEX);
+    }
+
+    @Test
+    public void execute_invalidIndexFilteredBookList_failure() {
```
Comment (qiu-siqi) [inline src/test/java/seedu/address/logic/commands/SelectCommandTest.java]: move above search result tests

File: src/test/java/seedu/address/logic/parser/SearchCommandParserTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import seedu.address.logic.commands.SearchCommand.SearchDescriptor;
+import seedu.address.testutil.SearchDescriptorBuilder;
+
+public class SearchCommandParserTest {
+    private static final String DEFAULT_SEARCH_TERM = "search term";
+
+    private SearchCommandParser parser = new SearchCommandParser();
+
+    @Test
+    public void parse_missingParts_failure() {
```
Comment (qiu-siqi) [inline src/test/java/seedu/address/logic/parser/SearchCommandParserTest.java]: parse_noFieldSpecified_failure()

File: src/test/java/seedu/address/logic/parser/SearchCommandParserTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        userInput = TITLE_DESC_ARTEMIS;
+        searchDescriptor = new SearchDescriptorBuilder().withTitle(VALID_TITLE_ARTEMIS).build();
+        expectedCommand = new SearchCommand(searchDescriptor);
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
+
+    @Test
+    public void parse_multipleRepeatedFields_acceptsLast() {
+        String userInput = DEFAULT_SEARCH_TERM + TITLE_DESC_ARTEMIS + TITLE_DESC_ARTEMIS + AUTHOR_DESC_ARTEMIS
+                + AUTHOR_DESC_ARTEMIS + TITLE_DESC_ARTEMIS + AUTHOR_DESC_ARTEMIS;
```
Comment (qiu-siqi) [inline src/test/java/seedu/address/logic/parser/SearchCommandParserTest.java]: need to use different books to check that last one is accepted...

File: src/test/java/systemtests/SearchCommandSystemTest.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        /* ----------------------------------- Perform invalid search operations ------------------------------------ */
+
+        /* Case: no search term or parameters -> rejected */
+        assertCommandFailure(SearchCommand.COMMAND_WORD, SearchCommand.MESSAGE_EMPTY_QUERY);
+
+        /* Case: no search term or parameters -> rejected */
+        assertCommandFailure("   " + SearchCommand.COMMAND_WORD + "             ", SearchCommand.MESSAGE_EMPTY_QUERY);
+
+        /* Case: mixed case command word -> rejected */
+        assertCommandFailure("SeaRcH hello", MESSAGE_UNKNOWN_COMMAND);
```
Comment (qiu-siqi) [inline src/test/java/systemtests/SearchCommandSystemTest.java]: test with "searchh"?

File: src/main/java/seedu/address/logic/commands/SearchCommand.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Search for books on Google Books\n"
+            + "Parameters: SEARCH_TERM "
+            + "[" + PREFIX_ISBN + "ISBN] "
+            + "[" + PREFIX_TITLE + "TITLE] "
+            + "[" + PREFIX_AUTHOR + "AUTHOR] "
+            + "[" + PREFIX_CATEGORY + "CATEGORY]\n"
+            + "Example: " + COMMAND_WORD
+            + PREFIX_TITLE + " Artemis "
+            + PREFIX_AUTHOR + "Andy Weir";
```
[the remark this answers is not in the document]
Comment (takuyakanbr) [reply src/main/java/seedu/address/logic/commands/SearchCommand.java]: Should we be providing multiple examples here? Might be less cluttered to keep them in the user guide?

File: src/test/java/seedu/address/logic/commands/SearchCommandTest.java (hunk 2 above)
Comment (takuyakanbr) [reply src/test/java/seedu/address/logic/commands/SearchCommandTest.java]: No it just returns an empty `SearchDescriptor`.

**Pull request merged**